src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala

```diff
@@ -0,0 +1,614 @@
+  * @param maxMismatches the maximum mismatches to match a sample barcode.
+  * @param minMismatchDelta the minimum difference between number of mismatches in the best and second best barcodes for
+  *                         a barcode to be considered a match.
+  * @param maxNoCalls the maximum number of no calls in the sample barcode bases allowed for matching.
+  */
+private class FastqDemultiplexer(val headers: Seq[SAMFileHeader],
+                                 val samples: Seq[Sample],
+                                 val readStructures: Seq[ReadStructure],
+                                 val umiTag: String = "RX",
+                                 val qualityFormat: FastqQualityFormat = FastqQualityFormat.Standard,
+                                 val minQ: Int = 0,
+                                 val maxQ: Int =  SAMUtils.MAX_PHRED_SCORE,
+                                 val maxMismatches: Int = 2,
+                                 val minMismatchDelta: Int = 1,
+                                 val maxNoCalls: Int = 2) {
+  import FastqDemultiplexer._
+
+  require(readStructures.nonEmpty, "No read structures were given")
+  require(samples.map(_.sampleOrdinal).toList == samples.map(_.sampleOrdinal).sorted, "Samples not provided in ascending order by sample ordinal")
+  require(samples.flatMap(_.sampleBarcode).map(_.concatenatedBarcode).sorted.distinct.length == samples.length, "Unique sample barcodes required for all samples")
+  require(headers.length == samples.length + 1, s"Not enough SAMFileHeaders; found ${headers.length} but should have ${samples.length+1}.")
+  require(headers.forall(_.getReadGroups.size == 1), "All headers must have one read group. Ordinals are: " + headers.zipWithIndex.filter(_._1.getReadGroups.size != 1).map(_._2).mkString(", "))
+
+  /** The number of reads that are expected to be given to the [[demultiplex()]]. */
+  def expectedNumberOfReads: Int = readStructures.length
+
+  /** True if we the read structure implies paired end reads will be produced, false otherwise. */
+  val pairedEnd: Boolean = readStructures.count(_.template.nonEmpty) match {
+    case 1 => false
+    case 2 => true
+    case n => throw new IllegalArgumentException(s"Found $n read structures with template bases but expected 1 or 2.")
+  }
+
+  /** The concatenation of all read structures read structure. */
+  private val fullReadStructure: ReadStructure = new ReadStructure(segs=readStructures.flatten, resetOffsets=true)
+
+  /** The quality format converter. */
+  private val solexaQualityConverter: SolexaQualityConverter = SolexaQualityConverter.getSingleton
+
+  /** The sub-reads for the sample barcodes across all read structures. */
+  private val sampleBarcodeSegments = fullReadStructure.sampleBarcode
+
+  /** Get the sample barcode bases across all bases. */
+  private def sampleBarcodeBases(allBases: String): String = {
+    ReadStructure.structureRead(bases=allBases, segments=sampleBarcodeSegments).map(_.bases).mkString
```

Comment:
I think I see why there are both, but it would be nice to note why you have this and also the one below.  Could even just be a single method that takes in the delimiter so you can use it both ways?